### PR TITLE
Add new music workshop

### PR DESCRIPTION
**Pull request by jkwok91**
*basically ready*

This is a PR to discuss the draft of a new workshop that introduces JS while playing with sound. Right now it's super lightweight and beginner-friendly (hopefully), but I'm hoping to ramp it up with a few slightly complex suggestions to keep people inspired.

### CODE REVIEW

**Comment by zachlatta** [comment]
Alright, awesome. I just went through it as a user. Feedback is as follows:

Quick overview of my thoughts:

- I think the first thing we need to do is sit down and figure out where this workshop will fall in the curriculum plan (is it the 2nd workshop? 3rd?). I think doing this will solve most of the problems around introducing concepts because we'll know what the user knows coming into the workshop. Mention me in a message in `#meta` on Slack so we can discuss.
- I think the workshop would benefit from having a more concrete purpose or thing you're building. Right now it feels like we're just throwing some buttons that play sounds onto the screen. I think it'd be much cooler if we built an "instrument" or specific buttons meant to be used to play a specific song. I suggested an idea for this below where I mentioned a YouTube video.

Complete thoughts follow:

#### _Preface_

- Why is there a horizontal rule at the very beginning of the workshop?
- The first question I have is "what is a soundboard"? Maybe show pictures and link to a YouTube video of someone using them?

#### Part I: Set Up

- I'd break down the Cloud9 setup into bullet points. Easier to read and follow.
- You should clarify to right click the root folder in the workspace. If they right click any other directory in Cloud9, they'll be creating their files for this workshop in the wrong directory.
- Remind people to save `index.html` after they type in the HTML template. Cloud9 doesn't automatically save by default.

#### Part II: Connect HTML and JS Files

- Introduction on linking HTML and JavaScript files can be improved. Does the beginner even know what JavaScript is yet?
  - I wouldn't spent too much time on this yet though. After we know where this workshop will fall in the curriculum plan (i.e. will it be the 2nd workshop? 3rd workshop?), we'll have a better idea of what the user will know going into it.
- It's unclear where to put the `script` tag in the HTML document. The user probably doesn't know what you mean by "the body of the HTML file" yet.
- Yeah, jQuery introduction can also be improved. I'd similarly not spend much time on improving this yet though, since we don't know what we can assume the user knows/doesn't know without knowing where this workshop falls in our curriculum plan.

#### Part III: User Interface

- The user hasn't viewed their blank page yet (`Preview > Live Preview File (index.html)`). I feel like we should at least have had them load it by now.

##### Adding Buttons

- What is the button tag? Might deserve some sort of introduction.
- I think we could use much more interesting sounds for this one. Is there a popular song out we could use sounds from?
  - Kind of off topic, but I found https://youtu.be/IhZ_o833NtI pretty entertaining on multiple levels. Try loading it and then hitting `9 8 7 9 8 5 7 9 8 7 8 9` on your keyboard to jump to different parts of the video
    - What if we extracted soundbites from each of those sections of the video and used those as sounds? We could also add symbols and stuff.
- Where do I put my code with the button tags? It's not super clear.
- Okay, I've added all the buttons you told me too, but I still have no idea what this looks like on the page. It may be worth having the user live preview their file by now.
- I think the `id` attribute deserves a better introduction

##### Adding ID's to Buttons

- It should be "Adding IDs to Buttons" or "Adding `id`s to Buttons"
- Yeah, I think this deserves a better introduction.
- The `-button` part of `cat-button` and `frog-button` is redundant.
  - Speaking of, I can totally see users trying to have spaces in their `id`s. Might be worth saying that you shouldn't have a space in an `id`.

#### Part IV: The JavaScript File

- I might not even worry about adding jQuery to the project until this section. It may make more sense to say something like "there's this fancy library called jQuery that makes it easy to trigger code when things on the page are clicked, so let's go ahead and add it"
- I'd replace all instances of the JS abbreviation with "JavaScript". Less terminology for the user to remember. And it's a healthy reminder that we're talking about JavaScript, not Java (speaking of, we should make sure to clarify this when/where we figure out how to properly introduce JavaScript for the first time. An analogy I like to use is "Java is to JavaScript as grapes are to grapefruit").

##### Playing the Sound

- We don't have `cat.mp3` uploaded anywhere yet, right?
- "Here, we've created a new `Audio` object that will have the sound `cow.mp3`..."
  - It's important for us to figure out and accommodate for how much the user knows going into this section. We're throwing an awful lot of terminology at them right now. What's an object? What's a variable? How did `cow.mp3` get into the javaScript? Where does that file come from? The user may not even know that code runs in order, line-by-line, yet. What's a method? Or a function?
- Code examples don't work for me because the MP3 files aren't uploaded anywhere yet.
- Before creating a cacophony of sounds, have the user load the page with just one sound playing so they know that at least something is working.

##### Creating Functions

- Like other sections of this workshop where terminology is introduced, I think we should rewrite this after we figure out what the user knows going into this workshop.
  - I think most of the wording trouble is a result of not knowing what assumptions we can (or can't) make yet
- `makeCatSound()` can have a better name. Perhaps `playCatSound()`?

##### Detecting the Click

- I'll just wait for you to finish this section before giving feedback. There's not much for me to talk about yet.

#### _Remainder of Project_

- Same as above. Just gonna hold off for now, since there's not much for me to give feedback on.
- Let me know if there are any particular things you're wrangling with, happy to discuss here or in `#meta` on Slack.

**Comment by ad510** [comment]
You wrote this 4 days ago? Sorry I missed it earlier!

Zach did a good job with general comments, so I'm just going to focus on how we introduce JS to someone who's programming for the first time.

There are lots of places where you mention programming concepts without really explaining what they are. For example:

- What's an "object"?
- When you write "A variable in Javascript is something that stores a value," what do you mean by "store" and what's a "value"?
- Why should you declare variables before using them?

On the other hand, we're trying to make the workshops focus on building an end product instead of being a theoretical intro to programming abstractions. So my first big picture question is how are we going to try to balance showing how to build something and having people understand the abstractions we're teaching? In other words we don't want people to have the illusion that they how to build stuff, then they try to make their own project and get lost on how to start (which often happens with things like Codecademy).

I don't know about you, but I suggest (and you can say whether you think it's a good idea) delaying introducing as many confusing abstractions as possible until they're actually needed, so that when we introduce them it makes sense what they're needed for. Maybe it's unclear what I mean so let me go on a rant to explain.

Rant about how we can introduce new concepts
--

First, a few notes:

- I tried explaining this on Slack at one point, but I did a bad job of explaining it, so I'll try to be clearer here.
- This rant is kind of abstract and theoretical, but I think most learn-to-code tutorials get this wrong and I think we have a good opportunity to not repeat their mistakes. (Every time we work on a workshop, I think we should keep in the back of our mind "why are we writing our own workshops when there's so many good learn-to-code websites we can point to instead?")
- Ideally I would lead by example by contributing my own workshops instead of ranting about my opinions about the curriculum arc, but as I learned from Bigfoot, it's time consuming to make workshops and even more time consuming to coordinate fitting them in the curriculum arc. Unfortunately for now that I don't have the time to do that so this is my next best option. And even if I did contribute more workshops, I'd have to explain my thoughts on the curriculum arc anyway in order to explain where my workshops fit in.

So here goes:

After we learn programming, we know about things like variables, functions, if statements, and looping constructs that we pretty much take for granted. But if you dump them all on someone learning them for the first time, they are *really* intimidating, and what's always on their mind is "Why do we need this? Why isn't there a simpler way to do this that I would understand better?"

Computers exist so that we can automate processes that we would otherwise have to do manually every single time. (Think of how much time that saves you compared to if we didn't have computers.) But the problem with computers is that computers are really dumb. In order to teach a computer how to do something, you have to break it down into tiny steps that it understands. But once you teach a computer how to do something once, it remembers how to do it forever, and you don't have to teach it how to do the same exact thing ever again.

Before my dad taught me programming, I tried to make Tic-Tac-Toe in a PowerPoint clone called Hyperstudio. Every possible board position would have its own slide, and every open board position would have a button pointing to a different slide. When my dad found out, he taught me how to make this in Visual Basic so I could teach it the rules of Tic Tac Toe once, so that the computer knew how to automatically create the board positions and I wouldn't have to manually repeat the same process to draw the board every single time.

So for a few years, the main thing I did with programming was making stupid games in Visual Basic. You know, draw some shapes in the IDE, add some buttons and a timer, then when an event happens, update the shapes using for loops. When I look back at my old code, one thing that stands out is for 3 years (!) after I first learned programming, I never used structures (you'd call them objects) or dynamically resized arrays, except in the occasional case where I copy-pasted someone else's code. I think I knew they existed (since I read other people's code online and in books), but apparently I didn't think I needed them. For example, instead of having an array of structs, why not just put each element in its own separate array? And the same was true for classes until I made Flash games, 5 years after I first learned programming! And I didn't really understand REST APIs until I made [Amelia](https://amelia-burbankgamehack.appspot.com) with Zach, 10 years after I first learned programming!

After you know the proper way to do these abstractions, it's easy to look back and say "that was dumb." And these things are indeed really useful, so obviously I think it's important that club members eventually learn them all. But when writing workshops, I think it's important to put yourself in the shoes of someone learning these things for the first time, who is intimidated by all these concepts and wondering "why do I need this and is there a simpler way?"

The answer, of course, is no, you don't need the abstractions, but if you don't use them then you have to do things manually yourself every single time. I think understanding what that really means is one of the most important things distinguishing hackers who build things from people who just program "for a job." And if you compare how self-taught hackers learn programming from how it's taught in schools, I suspect you'll find that self-taught hackers "naturally" learn programming by doing many new abstractions manually "the wrong way" at first, but by doing that they can better understand why "the right way" is better, and more generally have more appreciation for yet-undiscovered ways that computers can automate things.

How can we apply this to workshops? I think our curriculum should reflect how self-taught hackers learn things (the good parts of it, anyway), and I think it can be done something like this. For every new concept that can be potentially confusing, try to do the following:

1. Find a subset of the abstraction that can be done manually and still do something useful. This may take some creativity since this is usually considered "the wrong way" to do things, but I'm happy to help brainstorm.
2. Show how to do it manually. And don't worry that we're teaching how to do things "the wrong way" since (1) it'll be easier for people to understand, and (2) it'll still look magical to people when they're blissfully unaware (for now) that there's a better way to do it. (Think me not using dynamically sized arrays for 3 years.)
3. Explain how they are unnecessarily repeating themselves or reinventing an existing library or API.
4. Show the "right way" to do it. Now it's even more magical because you can do it in *that* much less code, and you can also do *this* other thing that would be super hard the old way!

The hardest part is step 1, a.k.a. seeing which concepts would actually be confusing to beginners then finding a clever way to delay teaching it. Once you get step 1 down, the rest is straightforward. So to see an example of doing things the "wrong way," it might help to view source of the [Bigfoot games](http://andrewd.50webs.com/bigfoot). The "proper way" to make an HTML5 game is to use a canvas-based library like Phaser, but making even a basic Phaser game requires knowing how to hook up APIs and knowing a lot of concepts up front. I think people who never programmed before would find that intimidating.

But instead, you can make a minimal Find Bigfoot game only using JS for the alert box, or even make a point-and-click adventure game with no JS at all by making images linking to other pages. Then you can add a timer and set .left and .top to move Bigfoot around. If the timer is short enough then you can store the current position as variables and move Bigfoot smoothly (reinventing the game loop). Then you can use keyboard events to move the player too. Then use this *really complicated* if statement with 4 parts to see if the player collided with Bigfoot (see this diagram showing how it works!), and increment the score. But wait there's more! You can even use createElement and appendChild to make new Bigfeet in JS. So now you use the timer to spawn new Bigfeet. Or use a for loop to create a bunch of Bigfeet *at the same time*. Or put the Bigfeet in an array so you can use the same code to animate dozens of Bigfeet. (Of the 3 games, only Bigfoot Defense uses arrays or loops.) And so on.

Oh, and by the way, see all those hard things you did? (Aren't you proud?) Some really smart people made this library called Phaser which does all those hard things and more for you. See, you can spawn new objects like this. And do collision detection like this. And you can do these fancy effects like rotation and gravity and so on that would be so much harder to do manually. Isn't that so much nicer? I bet in your life right now, there are things we are doing manually [without even knowing it](http://paulgraham.com/schlep.html) and you can find a way to make them easier like Phaser does and that will make everyone so much happier. Want to try? Have you heard of hackathons? 3 2 1 hack!

FYI, the Bigfoot games were made solely using Notepad, Paint.NET, Windows Explorer, and Chrome, and we didn't use any external libraries. Teaching to do things first the manual way basically means finding clever ways to avoid unnecessary dependencies, and a counterintuitive side effect of that is it delays teaching super useful things like JQuery or REST API calls. I completely agree that, say, REST APIs are super useful and should be an essential part of the curriculum. But every time you introduce a new concept or require making a new account, people are going to wonder "why." If they implement a small part of the functionality themselves first, it will be much easier to explain.

Also, the Bigfoot games was just an example and I'm not advocating basing our workshops on them. Maybe we can discuss further on Slack or Hack Night. Also for Hack Night I'm thinking of putting all my old Visual Basic programs online as an example of what self-taught beginner hacker code looks like.

I'm not going to advocate this approach religiously because I understand it's more important to complete workshops than to make it perfect. But I felt I had to explain it so you can understand my overall thinking and not just nitpicks on workshops you write.

End Rant
--

- BTW when you ask "what is a method," it's just a function attached to an object a certain way. Maybe just call it a "function in the Audio object" for now and avoid using the word "method" until later?
- Instead of using JQuery, why not just use plain JS instead which is often just as simple? I once saw a side by side comparison demonstrating that but I can't find it right now, and personally I just use plain JS without using JQuery in most my own code.

I hope to have more comments on the actual workshop, but for now I need some sleep. (I basically stayed up overnight to write this, oops.)

**Comment by ad510** [comment]
BTW I'm looking at a Phaser tutorial and it's not as bad in terms of JS concepts as I expected, so it likely does make sense to jump straight into a game library instead of making an HTML5 game manually. Though that's not the main point of my post.

**Comment by ad510** [comment]
I'm having second thoughts about some of the other things I wrote in the rant, but I'll need to sleep on it. I hope you find it useful for ideas even if parts of it are wrong or bad ideas. I'd been wanting to write approximately that rant for a while now.

**Comment by zachlatta** [comment]
Great rant, Andrew.

I think it can be summed up as:

   - We should teach people to program by helping them become self-taught
   programmers, not by introducing each concept upfront
      - What I mean by this, is that we should walk them through the same
      steps that a self-taught programmer would. Like doing things the
crappy way
      (ex. want to send 100 text messages in the Twilio workshop? copy
and paste
      the line to send a text message 100 times. later figure out you
can send a
      million text messages using a for loop)
   - If we just introduce each concept upfront, we're no different than any
   other learn to code resources, like Codecademy or a traditional computer
   science class. Hack Club is about *hacking*. Building things. Being
   thrown off the deep end and trying not to drown.
      - I'm a big fan of the idea of helping people build things and
      focusing on getting people excited, with the majority of learning taking
      place later when club members are working on projects and Google things.
      We've talked about this idea previously, but not in-depth when designing
      our new curriculum.

I'm in agreement with pretty much everything you said.

Instead of using JQuery, why not just use plain JS instead which is often
> just as simple? I once saw a side by side comparison demonstrating that but
> I can't find it right now, and personally I just use plain JS without using
> JQuery in most my own code.
>

I used to be in the same boat as you, but when you Google for basic DOM
manipulation stuff, it's almost always jQuery that you find. jQuery also
makes things like event listeners *way* easier.

```js
// jQuery
$("#myButton").click(function () {
  console.log("I've been clicked!");
});

// Standard JavaScript
var btn = document.getElementById("myButton");

myButton.addEventListener("click", function () {
  console.log("I've been clicked!");
}, false);
```

Yes, it could be simplified to
`document.getElementById("myButton").addEventListener...`, but that creates
a super long line with a lot going on.

On Fri, Apr 15, 2016 at 7:19 AM, Andrew Downing <notifications@github.com>
wrote:

> I'm having second thoughts about some of the other things I wrote in the
> rant, but I'll need to sleep on it. I hope you find it useful for ideas
> even if parts of it are wrong or bad ideas. I'd been wanting to write
> approximately that rant for a while now.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/hackclub/hackclub/pull/784#issuecomment-210480014>
>



-- 
Zach Latta
https://zachlatta.com <http://zachlatta.com>

**Comment by jkwok91** [comment]
Thanks for the thoughts, @ad510!! Really appreciate your feedback.
@zachlatta summed up what I was thinking, but a few additional points:
- I'm trying to do the whole "bumping into walls naturally and then rerouting / finding a better way" in the lines where I have each button declared in the same way. I was planning to have a section later on that condenses it, but currently am not sure about the jump in difficulty.
- Plain JS vs jQuery: I debated this, because it's a little confusing to see `$('#button').stuff(callback)`, but it might be quicker to explain than having to go through what `document` is, and after explaining all the DOM stuff, having to explain why the more common approach is to _not_ use plain JS and instead go for a library that does it for you.

**Comment by ad510** [comment]
OK, makes sense why you'd use jQuery.

The main thing I was having second thoughts on was when I think through how I'd actually introduce each of the programming concepts, it often doesn't make sense to literally walk through a hacky way to avoid using an abstraction then explain the abstraction, when it would be much simpler to just get to the point by explaining the abstraction. (Basically skip step 2 in the 4 steps I gave, since people can often figure out the hacky way to avoid abstractions by themselves.) But it's still useful to think through how someone would do something without knowing each particular abstraction, since often it's not necessary to use the abstraction right away and we want people to make cool things while introducing as few abstractions as possible. So my workshop thought process is really more like this:

1. Think of something cool we can teach how to make, and how we would make it.
2. For every abstraction that seems necessary to make it, think of if there's a hacky way to do it manually instead.
3. Teach how to make a simplified version of the thing, using as few abstractions as possible. When people build upon it by themselves afterwards, they can figure out the hacky ways to extend it without using the proper abstractions to approach what we originally thought of in step 1.
4. Repeat for something a little more complicated to make.

Versus what the textbooks do, which is more like this:

1. Think of all abstractions necessary to get a programming job.
2. Sort them from easiest to hardest.
3. For each new abstraction we want to teach, try to think of a few cool examples that use this abstraction and all previous abstractions we taught, but none of the other ones.
4. Explain how the abstraction works using those examples.
5. Repeat for the next abstraction.

In other words, we try to teach how to make something cool rather than teach what the programming abstractions are, and by doing it that way I think people will actually understand the abstractions better as a result.

I suppose this is something we intuitively know already, but I imagine it's a good idea to explicitly spell out how we're approaching things differently from other people, so we can keep that in the back of our mind when making workshops.

I added to my todo list to comment on the actual workshop more carefully, but for now I think Zach did a good job with that so I may not end up having much to add.

File: workshops/soundboard/README.md [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+---
+
+Today we're going to create our very own soundboard. You may have seen musicians such as Little Dragon, CHVRCHES, and [other musicians], using such things to improvise fresh beats at live performances. We can simulate this very easily with basic HTML and JavaScript.
+
+Much like in Personal Website workshop, we'll be using HTML to create stuff visible on the browser. We'll also be using JavaScript, a programming language commonly used in web development, to include interactivity in our project. In addition, we will be making use of jQuery, which is a library that will make using JavaScript in the browser much simpler. This may seem like an exorbitant number of new tools, but we'll be biting off manageable pieces.
+
+Table of Contents
+- Part I: Set Up
+- Part II: Connect HTML and JavaScript Files (question: should this be a subsection of Set Up)
```
Comment (MrCodeCollector) [inline workshops/soundboard/README.md]: Yes it should.

Comment (jkwok91) [inline workshops/soundboard/README.md]: Okay, idea for subsections. How y'all feel about this?
####Part I: Set Up
**Creating the Files**
**Adding JS to the HTML File**

File: workshops/soundboard/README.md [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+
+Short link to this workshop: [https://workshops.hackclub.com/soundboard](https://workshops.hackclub.com/soundboard)
+
+---
+
+Today we're going to create our very own soundboard. You may have seen musicians such as Little Dragon, CHVRCHES, and many other musicians, using such things to improvise fresh beats at live performances. We can simulate this very easily with basic HTML and JavaScript.
+
+Much like in Personal Website workshop, we'll be using HTML to create stuff visible on the browser. We'll also be using JavaScript, a programming language commonly used in web development, to include interactivity in our project. In addition, we will be making use of jQuery, which is a library that will make using JavaScript in the browser much simpler. This may seem like an exorbitant number of new tools, but we'll be biting off manageable pieces.
+
+Table of Contents
```
Comment (zachlatta) [inline workshops/soundboard/README.md]: Needs a newline after this line.

File: workshops/soundboard/README.md [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+We can create functions for each sound by wrapping the code we've already written in the following way:
+
+```js
+function makeChop3Sound() {
+  var chop3Sound = new Audio("chop3.mp3");
+  chop3Sound.play();
+}
+```
+
+(i need a better way to show that we are wrapping the two lines we already wrote in this function syntax)
```
Comment (zachlatta) [inline workshops/soundboard/README.md]: Should probably remove this line.

File: workshops/soundboard/README.md [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+    <html>
+      <head>
+      <title>Soundboard</title>
+    </head>
+    <body>
+    </body>
+    </html>
+  ```
+
+Similar to our HTML file in the Personal Website workshop, we've created a bare template upon which we will build interactions for the user. As you can see, we've named this page "Soundboard" using the title (`<title>`) tag.
```
Comment (jkwok91) [inline workshops/soundboard/README.md]: Markdown failure happens here; enumerated list suddenly dies.

File: workshops/soundboard/README.md [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+  if (event.keyCode == 100) {
+    var chop3Sound = new Audio("chop3.mp3");
+    chop3Sound.play();
+  }
+}
+``
+
+Simple enough, we just reuse our code from earlier.
+
+But wait, we already have preexisting code to play the chop3 sound. It's in `playSound()`, and chop3 will be played when the chop3 button is clicked (à la `$("button").on("click",playSound);`)! If only we could somehow use that for the chop3 button.
```
Comment (zachlatta) [inline workshops/soundboard/README.md]: What is up with that apostrophe?

File: workshops/soundboard/README.md [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+  <button>Kick4</button>
+  <button>Perc1</button>
+  <button>Vox9</button>
+  <script src="https://code.jquery.com/jquery-2.2.3.min.js"></script>
+  <script src="main.js"></script>
+</body>
+</html>
+```
+
+Now, as a user, we can tell the difference between each of the buttons, by what text is displayed on the buttons. But how can we differentiate them in code? One easy way is to supply each with an attribute known as an "`id`."
```
Comment (jkwok91) [inline workshops/soundboard/README.md]: inconsistency with the use of the term "id" (sometimes in `code` and sometimes not)

File: workshops/soundboard/README.md (hunk 4 above)
Comment (zachlatta) [inline workshops/soundboard/README.md]: Gotta indent the stuff properly.

**Comment by zachlatta** [comment]
We decided to go ahead and merge this. We'll make sure to improve it in the coming days – there are tons of things that need improvements.

**Comment by zachlatta** [comment]
:tada: :tada: :tada:
